### PR TITLE
Bounds parsing refactoring 2

### PR DESCRIPTION
**Pull request by petrochenkov**
See https://github.com/rust-lang/rust/pull/37511 for previous discussion.
cc @matklad

Relaxed parsing rules:
 - zero bounds after `:` are allowed in all contexts.
 - zero predicates are allowed after `where`.
- trailing separator `,` is allowed after predicates in `where` clauses not followed by `{`.

Other parsing rules:
 - trailing separator `+` is still allowed in all bound lists.

Code is also cleaned up and tests added.

I haven't touched parsing of trait object types yet, I'll do it later.

### CODE REVIEW

**Comment by durka** [comment]
Why disallow trailing `+` when trailing `,` is allowed? Does it create an ambiguity?

**Comment by petrochenkov** [comment]
@durka 
No ambiguity, it's a [lang team decision](https://github.com/rust-lang/rust/pull/37511#issuecomment-264312414).

(I'd personally keep it though, because 1) uniformity (other trailing separators are allowed), 2) it was accepted since 1.0 until now, 3) it's "naturally" allowed and you have to add extra code to the parser to prohibit it)

**Comment by petrochenkov** [comment]
@durka 
Do you have an example of trailing `+` being useful in macros?

**Comment by durka** [comment]
Yeah, those three reasons make sense to me. This should be cratered then,
WRT #2.

**Comment by durka** [comment]
I don't think it's useful in macros, because you can just do `$($bounds)++`
and it's no problem. I was thinking of consistency between various types of
separators. But then again you don't want "1 + 1 +" to be a valid
arithmetic expression.

**Comment by petrochenkov** [comment]
>But then again you don't want "1 + 1 +" to be a valid
arithmetic expression.

Expression `+` and type `+` are actually different! :slightly_smiling_face: 
Expression `+` is a binary operator and lists like `1 + 1 + 1 + 1` are desugared into several operator applications `((1 + 1) + 1) + 1`.
Type `+` is not a binary operator and summands in lists like `Trait + Trait + Trait` can't be grouped. `(Trait + Trait) + Trait` and `ArbitraryType + Trait` in general are not valid.

**Comment by nikomatsakis** [comment]
@petrochenkov I feel like I would keep it too -- which is probably why it's parsed in the first place. I wonder if it's worth revisiting this. I found your three points convincing, and I'm not sure that we focus on the backwards compatibility point in the meeting.

cc @rust-lang/lang -- we had [previously decided](https://github.com/rust-lang/rust/pull/37511#issuecomment-264312414) to not accept `T: Foo + Bar +` (where `+` acts as a trailing separator) because it "looks weird" (which is indisputably true). However, @petrochenkov points out that it [is accepted today](https://is.gd/th9sfI) (also indisputably true). "Looking weird" doesn't feel like reason enough to break backwards compatibility to me, particularly since accepting trailing separators is known to be useful for macros. So I think I'd like to revisit this decision. What do people think?

**Comment by withoutboats** [comment]
I think when we agreed on that we thought we were talking about expanding what is accepted to include trailing `+`. I agree we shouldn't break the language.

**Comment by joshtriplett** [comment]
@nikomatsakis If we accept this today, we should continue doing so, and document it for potential use by macros.  Not worth breaking backward compatibility over.

**Comment by aturon** [comment]
@nikomatsakis I agree with your reasoning; let's keep it.

**Comment by nikomatsakis** [comment]
@petrochenkov care to update the PR to accept trailing `+`?

@nrc you were the main one who objected, I think, so I'd like to get your input in particular

(Does anyone feel we need to use rfcbot here?)

**Comment by nrc** [comment]
@nikomatsakis yeah, I agree its not worth breaking for.

**Comment by petrochenkov** [comment]
Updated with trailing `+`s permitted.

**Comment by nikomatsakis** [comment]
@bors r+

Thanks @petrochenkov !

**Comment by nikomatsakis** [comment]
Oops :) had one thing I wanted to ask

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/librustc_passes/ast_validation.rs [hunk 1]
```diff
@@ -284,6 +290,26 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
...
+            if ty_param.default.is_some() {
+                seen_default = Some(ty_param.span);
+            } else if let Some(span) = seen_default {
+                self.err_handler()
+                    .span_err(span, "type parameters with a default must be trailing");
+                break
+            }
+        }
+        for predicate in &g.where_clause.predicates {
+            if let WherePredicate::EqPredicate(ref predicate) = *predicate {
```
Comment (nikomatsakis) [inline src/librustc_passes/ast_validation.rs]: to be clear, not as part of this PR =)

Comment (petrochenkov) [inline src/librustc_passes/ast_validation.rs]: There are no plans to implement this?
Then https://github.com/rust-lang/rust/issues/20041 should be officially closed or something before removing `WherePredicate::EqPredicate`.

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -1353,13 +1352,15 @@ impl<'a> Parser<'a> {
             self.expect(&token::CloseDelim(token::Bracket))?;
             t
         } else if self.check(&token::BinOp(token::And)) ||
-                  self.token == token::AndAnd {
+                  self.check(&token::AndAnd) {
             // BORROWED POINTER
             self.expect_and()?;
             self.parse_borrowed_pointee()?
         } else if self.check_keyword(keywords::For) {
+            // FIXME plus priority
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: `+` still has high priority in some cases.

The first case is when trait object type starts with `for`:
```
trait Tr<'a> {}
fn f(arg: &for<'a> Tr<'a> + Send){} // Parsed as &(for<'a> Tr<'a> + Send)
```
I'm going to fix this in upcoming PR so I didn't create an issue.

The second case is `impl Trait1 + Trait2` and it seems to be [known](https://github.com/rust-lang/rust/issues/34511#issue-162545419).

Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Updated with better comments for FIXMEs.

File: src/librustc_passes/ast_validation.rs (hunk 1 above)
Comment (nikomatsakis) [inline src/librustc_passes/ast_validation.rs]: Well, eventually perhaps, but at the moment there are some theoretical limitations that I do not yet know how to overcome. I would say "we may implement it eventually, but in that case we can add that stuff back in to the compiler".

**Review by nikomatsakis — APPROVED** [review]

**Comment by alexcrichton** [comment]
@bors: retry

* android emulator wedged? Haven't seen that before...

**Comment by alexcrichton** [comment]
@bors: retry

* fiddling with homu

**Pull request merged**